packages/contracts/test/voting/admin-address-setup.ts

```diff
@@ -0,0 +1,184 @@
+      );
+    });
+
+    it('fails if data is empty, or not of minimum length', async () => {
+      await expect(
+        adminAddressSetup.prepareInstallation(targetDao.address, EMPTY_DATA)
+      ).to.be.reverted;
+
+      await expect(
+        adminAddressSetup.prepareInstallation(
+          targetDao.address,
+          minimum_data.substring(0, minimum_data.length - 1)
+        )
+      ).to.be.reverted;
+    });
+
+    it('reverts if encoded address in `_data` is zero', async () => {
+      const dataWithAddressZero = abiCoder.encode(['address'], [AddressZero]);
+
+      await expect(
+        adminAddressSetup.prepareInstallation(
+          targetDao.address,
+          dataWithAddressZero
+        )
+      ).to.be.revertedWith(customError('AdminAddressInvalid', AddressZero));
+    });
+
+    it('correctly returns plugin, helpers and permissions', async () => {
+      const nonce = await ethers.provider.getTransactionCount(
+        adminAddressSetup.address
+      );
+      const anticipatedPluginAddress = ethers.utils.getContractAddress({
+        from: adminAddressSetup.address,
+        nonce,
+      });
+
+      const {plugin, helpers, permissions} =
+        await adminAddressSetup.callStatic.prepareInstallation(
+          targetDao.address,
+          minimum_data
+        );
+
+      expect(plugin).to.be.equal(anticipatedPluginAddress);
+      expect(helpers.length).to.be.equal(0);
+      expect(permissions.length).to.be.equal(2);
+      expect(permissions).to.deep.equal([
+        [
+          Op.Grant,
+          plugin,
+          ownerAddress,
+          AddressZero,
+          ADMIN_EXECUTE_PERMISSION_ID,
+        ],
+        [
+          Op.Grant,
+          targetDao.address,
+          plugin,
+          AddressZero,
+          EXECUTE_PERMISSION_ID,
+        ],
+      ]);
+    });
+
+    it('correctly sets up the plugin', async () => {
+      const daoAddress = targetDao.address;
+
+      const nonce = await ethers.provider.getTransactionCount(
+        adminAddressSetup.address
+      );
+      const anticipatedPluginAddress = ethers.utils.getContractAddress({
+        from: adminAddressSetup.address,
+        nonce,
+      });
+
+      await adminAddressSetup.prepareInstallation(daoAddress, minimum_data);
+
+      const factory = await ethers.getContractFactory('AdminAddress');
+      const adminAddressContract = factory.attach(anticipatedPluginAddress);
+
+      expect(await adminAddressContract.getDAO()).to.be.equal(daoAddress);
+    });
+  });
+
+  describe('prepareUninstallation', async () => {
+    it('correctly returns prepare uninstallation data abi', async () => {
+      // Human-Readable Abi of data param of `prepareUninstallation`.
+      const dataHRABI = '';
+
+      expect(await adminAddressSetup.prepareUninstallationDataABI()).to.be.eq(
+        dataHRABI
+      );
+    });
+
+    it('correctly returns permissions', async () => {
+      const plugin = ethers.Wallet.createRandom().address;
+
```

Comment:
Why don't we also test that the uninstall does the work?

Reply:
It is the work,
describe = 'prepareUninstallation'
it1 = 'correctly returns prepare uninstallation data abi'
it2 = 'correctly returns permissions'
it2 only returns permission and no more

i can rename it2 to "correctly prepares uninstallation" ,, WDYT?